### ISSUE TITLE
o365 default calendar broken (but not others) on 2023.4

### ISSUE THREAD

**Issue by robertoleonardo** [RogerSelwyn/O365-HomeAssistant#132]
Hi - this may be more an issue with homeassistant itself, but it's not clear to me because it's an issue only with 1 of my 3 o365 calendars....so i'll start here. 

o365 integration works great on 2023.3 with all of my work account calendars (default ["calendar"], holidays, and birthdays).  on 2023.4, the default calendar no longer loads on the HASS calendar page, or in atomic calendar revive.  the other 2 calendars still work (as do my calendars from other accounts).  and i created a new work account calendar and it also works.  so it's ONLY the default calendar that no longer works.  i've restored from backup to 2023.3 like 5 times to confirm its not something on he server side -- every time, the calendar immediately returns after restore.  so it's definitely related to 2023.4.  

the weird thing, though: the entity exists and it is getting data -- if i go to the developer tools and look at the state, it shows my next few events even from that calendar.  so it's not something about the connection, but it's also not clear to me that it's something wrong with HASS itself.  

nonetheless, although it's receiving data - this is what i see when i click 'calendar' in the sidebar:  
![image](https://user-images.githubusercontent.com/13153043/230541034-58971c14-f492-4791-a159-d7e82663c303.png)

again, the other calendars load with no issue.  it's just this single one that gives that error, even though it shows up in the left hand panel.  maybe the data for that particular calendar--for some reason--is getting formatted weirdly such that although the data is flowing to home assistant, none of the calendar cards/integrations can handle it? idk. 

i'm having trouble finding anything relevant in logs -- but if you are able to point me in the right direction re logs -- i'll send you whatever i can to help.

**Comment by RogerSelwyn** [comment]
Hi, this may be a difficult one to track down. There has been much work done on the HA core around calendar processing in recent months. An article was published recently around some expectations which may give some clues and which the module may not be 100% compliant to as yet. See below. In terms of logs, I would look for any errors at all in your logs, that said I tend to find those generated by the frontend to often give no clue as to what the problem is. In particular look at any logs generated when you first visit the calendar page and it try's to do a retrieval (it doesn't use the content you see in developer states, it does a separate call when you open page).

My first suspicion is that it is to do with all day events. To diagnose, you could, create a new calendar and turn off tracking on your default one, then copy across events from the default one to the new one (in batches) until you get the issue. If you are using MS Outlook this should be a simple task to do by changing the calendar view to a list.

I am away from Monday for a week and have limited time before then, so be prepared for living with this for a while.

https://developers.home-assistant.io/blog/2023/03/28/calendar_best_practices/

**Comment by robertoleonardo** [comment]
fair enough! and i'm buried with work for a while, in any case. i will dig into it, but won't get around to it for a few days or a week.  for the time being, i'm just living on 2023.3. 

thanks for the good suggestions and the FANTASTIC integration.  you're the man. 

ill update you once i've gotten to dig further.

**Comment by RogerSelwyn** [comment]
Just a quick check, what timezone are you in?

**Comment by RogerSelwyn** [comment]
Doing some changes and making the integration compliant to the link I added I accidentally noted that if you have an empty subject for an event, it silently fails within the core code. No idea why, but I'll try to add something that inserts an empty value rather than a None value. Maybe worth checking if you have any empty subjects.

**Comment by robertoleonardo** [comment]
pacific. i was wondering if that had something  to do with it because i had seen the calendar related comments in the 2023.4 release notes and several of them seemed to focus on utc time zone stuff.

i think your all day event point may be on to something -- but it seems like it might be the events with times, not all day events, that are problematic. because the two calendars that still work only have all day events. and when i created a new calendar before my first post, the two events i copied over to test were all day events. and it still worked.  i'll experiment further when i dig myself out from under work mountain and see if it's events with start and end times that are problematic.



--
rob lee
m: 310.431.1619
________________________________

**Comment by RogerSelwyn** [comment]
I think it is likely the problem is related to my last post about an empty Subject, because I have seen that cause exactly the symptoms you are showing in my testing today. I may publish a release before I go away with some code to capture the errors being generated by core, but equally may wait until after I get back.

**Comment by robertoleonardo** [comment]
ah interesting.  well I’ll be on the lookout for that release, whenever it may be – take your time and I hope you’re going away for fun purposes!  and if I find time to tinker before you get a release out, I’ll keep you posted on what I find.

**Comment by robertoleonardo** [comment]
looks like it's fixed in 2023.4.1! wooo! thanks for the help, roger

**Comment by robertoleonardo** [comment]
oh, interesting - spoke too soon.  it's only halfway fixed -- so, on 2023.4.2 -- i'm still getting the error when i click the built in calendar in the sidebar.  but now, suddenly with this new version, atomic calendar on my default dashboard works again.  

i'm also now seeing this in the log:  

![image](https://user-images.githubusercontent.com/13153043/230759233-dd26319e-c1ab-4b03-bbd1-2a638f3463a8.png)

**Comment by RogerSelwyn** [comment]
I haven't changed anything in 2023.4.1, not my code in the core.

That error suggests that the calendar no longer exists. Lookup the long alphanumeric in that url in your calendars.yaml, and make sure it exists. If it doesn't, then you can just delete from calendars.yaml.

**Comment by robertoleonardo** [comment]
ah yea, that log entry was irrelevant to the issue.  i just had an old, redundant yaml file in there somehow from all the restoring from backups i've been doing.  cleaned it up and the now the log is back to being entirely unhelpful with the real issue. oh well, i will keep tinkering. safe travels

**Comment by brunicholson** [comment]
I'm experiencing the same. Only default calendar is broken...

**Comment by robertoleonardo** [comment]
@RogerSelwyn - i did some tinkering with creating a new calendar and starting to copy events over and couldn't manage to break  the new calendar.  i didn't copy every last event over, but i tried several events of each type i could think of (all day - multi day; all day - single day; start/end single day; start/end multi day).  the test calendar continued to work fine.  given @brunicholson is also experiencing the issue and only with a default calendar -- i'm wondering if it has something to do with default calendars in particular, and not necessarily the event type.

**Comment by robertoleonardo** [comment]
Yep - see my last post. didn't fix it but i did notice a quirk that may or may not be relevant




--
rob lee
m: 310.431.1619
________________________________

**Comment by robertoleonardo** [comment]
Well, as a theoretical matter i'm with you on mandatory subject lines! But it doesn't seem to be the issue here, unless i'm misunderstanding you. I added events with blank subject lines to my test calendar and they did in fact show up without issue. The only thing that was a bit odd is that they did seem to prevent the test CALENDAR (not just the event) from appearing in HA for the first time. i created the test cal and immediately added a blank event and made sure the calendar had synced back to o365, only at that point did i restart HA. after the reboot,  no test calendar. so i deleted the blank event and rebooted again. then the test cal appeared.  im not entirely convinced the blank event was the issue -- might have just been a slight delay or something. but even if it was the issue on that first time loading the calendar, i got curious and added another several blank events to the test calendar now that it had showed up in HA. this time, the blank events showed up in ha immediately and i didn't get any errors on that. i rebooted a couple times and it continued to load without any issues.


and coming at it from the opposite angle --even after i deleted the single blank event on my DEFAULT calendar that  i was able to find (using filters with "subject is empty" as the only constraint -- so presumably it's the only one.), the issue didn't go away.

So perhaps it somehow relates to blank events given the quirk i noticed with my test calendar, but that doesn't seem to be the sole or direct cause -- because if so, you'd think that my default calendar would've loaded without issue after i got rid of the empty event.  but nope -- problem persists.

anyway, i'm only responding because i'm bored of working, and this issue has made i'm curious -- but don't worry, i am not losing sleep over this.  i super appreciate your thoughts.  seriously no rush here on my account. please feel free to ignore me/ this issue for as long as you want, and certainly at least until you're back in the country! safe travels



--
rob lee
m: 310.431.1619
________________________________

**Comment by RogerSelwyn** [comment]
Thanks for the info, I’ll investigate more when I get back. I did try a default calendar when you first reported it and it worked fine for me. I think if it was caused by the fact of being a default calendar I’d have loads of people reporting. In any case I’ll be able to post a alpha test build daily quickly to see if I can capture what every error is being reported back from core because I’m pretty sure it is something they did in the change from 2023.3 to 2023.4. But couldn’t find it obviously in the change log.

**Comment by robertoleonardo** [comment]
@RogerSelwyn - whenever you get around to it, it looks like google calendar people had the same issue: https://github.com/home-assistant/core/issues/90952

the issue seems to have been fixed with gcal at least. 

note that one of the comments in there says that the issue was limited to his primay/default calendar. interesting.

**Comment by robertoleonardo** [comment]
ah ha! they just released 2023.4.3 and it's fixed!

from the release notes -- i suspect this is what did it:  

Fix all day event coercion logic ([@allenporter](https://github.com/allenporter) - [#91169](https://github.com/home-assistant/core/pull/91169)) ([calendar docs](https://www.home-assistant.io/integrations/calendar/)) ([local_calendar docs](https://www.home-assistant.io/integrations/local_calendar/))

i'm going to leave this open for a few hours in case i end up wrong that it's fixed, bu t i think we're good to go now.

**Comment by robertoleonardo** [comment]
confirmed - the issue seems to be all fixed for me! wahoo.  i'll leave the issue open as i guess i don't definitively know that it's solved for @brunicholson -- but as far as i'm concerned, this issue is closed.  thanks @RogerSelwyn !

**Comment by brunicholson** [comment]
It has also solved it for me guys. Panic over 💯
